remill/BC/DeadStoreEliminator.cpp

```diff
@@ -0,0 +1,768 @@
+      return true;
+    }
+  }
+  return false;
+}
+
+bool LiveSetBlockVisitor::VisitBlock(llvm::BasicBlock *B) {
+  LiveSet live;
+  for (auto inst_it = B->rbegin(); inst_it != B->rend(); ++inst_it) {
+    auto inst = &*inst_it;
+    if (llvm::isa<llvm::ReturnInst>(inst) || llvm::isa<llvm::UnreachableInst>(inst) ||
+        llvm::isa<llvm::IndirectBrInst>(inst) || llvm::isa<llvm::ResumeInst>(inst) ||
+        llvm::isa<llvm::CatchSwitchInst>(inst) || llvm::isa<llvm::CatchReturnInst>(inst) ||
+        llvm::isa<llvm::CleanupReturnInst>(inst)) {
+      // code that we return to or branch to could read out registers
+      // so mark as all live
+      live.set();
+    } else if (llvm::isa<llvm::BranchInst>(inst) || llvm::isa<llvm::SwitchInst>(inst)) {
+      // update live from successors
+      for (llvm::BasicBlock *succ : successors(B)) {
+        // all successors must be default have been visited first
+        // b = U { s | for s in successors(b)}
+        live |= block_map[succ];
+      }
+    } else if (auto call_inst = llvm::dyn_cast<llvm::CallInst>(inst)) {
+      uint64_t slotn = 0;
+      // Alternative 1:
+      // mark as all live if called function has the same type as lifted function
+      // we're not able to see inside other paths, so we can't predict whether or
+      // not the callee might not use any of our slots
+      // Alternative 2:
+      // mark live based on how much of the state struct could be accessed
+      if ((call_inst->getFunctionType() == lifted_func_ty) || 
+          CallAccessesState(call_inst, &slotn)) {
+        if (!slotn) {
+          live.set();
+        } else {
+          live.set(slotn);
+        }
+      }
+    } else if (llvm::isa<llvm::InvokeInst>(inst)) {
+      if (call_inst->getFunctionType() == lifted_func_ty) {
+        // mark as all live if invoked function has the same type as lifted function
+        // we're not able to see inside other paths, so we can't predict whether or
+        // not the callee might not use any of our slots
+        live.set();
+      }
+    } else if (auto store_inst = llvm::dyn_cast<llvm::StoreInst>(inst)) {
+      if (auto scope = GetScopeFromInst(*inst)) {
+        auto inst_size = dl->getTypeAllocSize(store_inst->getOperand(0)->getType());
+        auto &state_slot = state_slots[scope_to_offset.at(scope)];
+        if (!live.test(state_slot.i)) {
+          if (on_remove_pass) {
+            to_remove.insert(inst);
+          }
+        } else if (inst_size < state_slot.size) {
+          // Partial stores revive the slot.
+          live.set(state_slot.i);
+        } else {
+          // Full stores kill the slot.
+          live.reset(state_slot.i);
+        }
+        // Mark that we have accessed this slot
+        func_used.set(state_slot.i);
+      }
+    } else if (llvm::isa<llvm::LoadInst>(inst)) {
+      if (auto scope = GetScopeFromInst(*inst)) {
+        // Loads revive the slot.
+        live.set(state_slots[scope_to_offset.at(scope)].i);
```

Comment:
Pull out the `state_slots[scope_to_offset.at(scope)].i` into a local var so that you don't need to do it twice.